resources/aws_ec2.go

```diff
@@ -525,11 +526,16 @@ func (obj *AwsEc2Res) snsWatch() error {
 		}
 	}()
 	obj.wg.Add(1)
-	// start the endpoint
+	// create the sns listener
+	listener, err := obj.snsListener(obj.WatchListenAddr)
```

Comment:
Out of curiosity: if you call shutdown and there's no listener, it doesn't panic, does it? (If it's easy to test, please do.)

Reply:
Also, I think this needs to happen *before* the server creation stuff, and also it needs to defer some sort of close https://golang.org/pkg/net/#Listener ...

That way we (1) open listener, (defer listener close (4)) (2) open server (defer server close (3))

Makes sense?

Reply:
If we close the listener manually, doesn't that defeat the purpose of the graceful shutdown? Server.Shutdown handles this for us, afaict. https://golang.org/src/net/http/server.go?s=47634:47680#L2492

Also Server.Shutdown(ctx) panics if the server isn't serving (ie. before running Server.Serve(listener)) so we'll have to make sure it is running before calling Shutdown(ctx). Not sure how to check that, though.

Reply:
Also, I think as long as it happens before snsServer.Serve(listener) we're ok, as the server creation just uses an addr, and the listener doesn't come into play until we start the server.

Reply:
Cool-- okay, just leave a comment in the source saying that the listener will get close by shutdown... (if we *never* serve) will the shutdown still close without panicing?

Reply:
Unfortunately it does panic. :( I looked at the server type and it wasn't obvious how to check if it's serving. I suppose we could make a HTTP request and shutdown if it doesn't error, but that seems kind of sloppy.